eth/catalyst/simulated_beacon_api.go

```diff
@@ -18,44 +18,82 @@ package catalyst
 
 import (
 	"context"
-	"time"
 
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/core"
 	"github.com/ethereum/go-ethereum/core/types"
-	"github.com/ethereum/go-ethereum/log"
 )
 
-type api struct {
+// simulatedBeaconAPI provides a RPC API for SimulatedBeacon.
+type simulatedBeaconAPI struct {
 	sim *SimulatedBeacon
 }
 
-func (a *api) loop() {
+// newSimulatedBeaconAPI returns an instance of simulatedBeaconAPI with a
+// buffered commit channel. If period is zero, it starts a goroutine to handle
+// new tx events.
+func newSimulatedBeaconAPI(sim *SimulatedBeacon) *simulatedBeaconAPI {
+	api := &simulatedBeaconAPI{sim: sim}
+	if sim.period == 0 {
+		// mine on demand if period is set to 0
+		go api.loop()
+	}
+	return api
+}
+
+// loop is the main loop for the API when it's running in period = 0 mode. It
+// ensures that block production is triggered as soon as a new withdrawal or
+// transaction is received.
+func (a *simulatedBeaconAPI) loop() {
 	var (
-		newTxs = make(chan core.NewTxsEvent)
-		sub    = a.sim.eth.TxPool().SubscribeTransactions(newTxs, true)
+		newTxs    = make(chan core.NewTxsEvent)
+		newWxs    = make(chan newWithdrawalsEvent)
+		newTxsSub = a.sim.eth.TxPool().SubscribeTransactions(newTxs, true)
+		newWxsSub = a.sim.withdrawals.subscribe(newWxs)
+		doCommit  = make(chan struct{}, 1)
 	)
-	defer sub.Unsubscribe()
+	defer newTxsSub.Unsubscribe()
+	defer newWxsSub.Unsubscribe()
+
+	// commit is a non-blocking method to initate Commit() on the simulator.
+	commit := func() {
+		select {
+		case doCommit <- struct{}{}:
+		default:
+		}
+	}
+	// a background thread which signals to the simulator when to commit
+	// based on messages over doCommit.
+	go func() {
+		for _ = range doCommit {
+			a.sim.Commit()
+			a.sim.eth.TxPool().Sync()
+			executable, _ := a.sim.eth.TxPool().Stats()
+			if executable != 0 {
+				commit()
+			}
+		}
```

Comment:
I don't love this loop. It is un-untuitively recursive, passing a struct over the chan that we're iterating on. It works due to chan buffering and non-blocking send, but can't we just do something more obvious, like for example: 

```golang
		for _ = range doCommit {
			a.sim.Commit()
			a.sim.eth.TxPool().Sync()
			for {
				a.sim.Commit()
				if executable, _ := a.sim.eth.TxPool().Stats(); executable == 0 {
					break
				}
			}
		}
```
This is _nearly_ the same, but not quite: it avoids sending on the channel we're iterating on, but OTOH we won'd notice if the channel becomes closed. 

In both cases, if for some reason we have an executable transaction which is never included in a block, this PR will spinloop here. 

And yeah, the code I pushed here is a bit unsafe: I close the `doCommit` in the outer loop, on exit, but the current code sends on it. So possibly send-on-closed crash here (my bad!).

Reply:
Ok, removed the recursion and the spin-looping, and the "send on closed". Now it does max one extra Commit on each loop, but not indefinite.

Reply:
Hmm the problem is that doing one extra commit per loop doesn't really fix the issue when the pending txs cannot be included over two blocks? It's an improvement over 1, but it can still leave txs in the pool.

Reply:
Yeah you were right, I had to change that

Reply:
The inner loop does make things much clearer though.